### PR TITLE
APS-602: Show Completed tasks in Task allocation section

### PR DESCRIPTION
**Pull request by vubalasu**
# Context

# Changes in this PR
Show Completed tasks in Task allocation section
### Before
Screenshot 2024-04-11 at 12 08 19

### After
Screenshot 2024-04-12 at 12 01 51

### CODE REVIEW

File: server/utils/tasks/listTable.ts [hunk 1]
```diff
@@ -28,6 +29,32 @@ const getTaskType = (task: Task): string => {
   return taskType
 }
 
+const getDecisionOutcome = (outcome: string): string => {
```
Comment (pezholio) [inline server/utils/tasks/listTable.ts:32]: I think I'd rather have `getDecisionOutcome` functions for both decision type, with either `AssessmentDecision`, `PlacementApplicationDecision` or `PlacementRequestTaskOutcome` passed in. That way you can have a record type like (for example):

```typescript
const assessmentDecisionTranslations: Record<AssessmentDecision, string> = {
    accepted: 'Accepted',
    rejected: 'Rejected'
}
```
 
And then you can have a getDecisionOutcome translator like so:

```typescript
const getDecisionOutcome = (outcome: AssessmentDecision): string => {
    return assessmentDecisionTranslations[outcome]
}
```

That way if any of the types change, we're alerted to this and can fix it.

File: server/utils/tasks/listTable.ts [hunk 2]
```diff
@@ -28,6 +29,32 @@ const getTaskType = (task: Task): string => {
...
+    case 'withdraw':
+      return 'Withdraw'
+    case 'withdrawn_by_pp':
+      return 'Withdrawn by pp'
+    default:
+      return ''
+  }
+}
+
+export const getDecision = (task: Task): string => {
```
Comment (pezholio) [inline server/utils/tasks/listTable.ts:47]: Looks like we're missing a case for `PlacementRequestTask`s?

**Review by pezholio — CHANGES_REQUESTED** [review]

File: server/utils/tasks/listTable.ts (hunk 2 above)
Comment (vubalasu) [reply server/utils/tasks/listTable.ts:47]: but we are requesting only 'PlacementApplication', 'Assessment'

Comment (pezholio) [reply server/utils/tasks/listTable.ts:47]: I think it'd be best to have it there for completeness, as this part of the code doesn't have any knowledge of what we've requested at the controller level. If this changes for any reason, or we try and use this code in a place where we request all types, we could end up with unintended consequences.

File: server/utils/tasks/listTable.ts [hunk 3]
```diff
@@ -93,13 +147,44 @@ const unallocatedTableRows = (tasks: Array<Task>): Array<TableRow> => {
...
 
-const tasksTableRows = (tasks: Array<Task>, allocatedFilter: string): Array<TableRow> => {
-  const rows: Array<TableRow> =
-    allocatedFilter === 'allocated' ? allocatedTableRows(tasks) : unallocatedTableRows(tasks)
+export const completedAtDateCell = (item: Task): TableCell => ({
+  text: item.outcomeRecordedAt ? DateFormats.isoDateTimeToUIDateTime(item.outcomeRecordedAt) : '',
+})
+
+export const completedByCell = (task: Task): TableCell => ({
+  text: task.allocatedToStaffMember?.name,
```
Comment (richpjames) [inline server/utils/tasks/listTable.ts:155]: If the `task` doesn't have an `allocatedToStaffMember` property the `text` property in the returned object will be `undefined`. It's better to return an empty string

File: integration_tests/mockApis/tasks.ts [hunk 4]
```diff
@@ -26,6 +27,7 @@ export default {
     sortBy: TaskSortField
     apAreaId: string
     types: Array<TaskType>
+    isCompleted: string
```
Comment (richpjames) [inline integration_tests/mockApis/tasks.ts:30]: I *think* this should be a `boolean`

File: server/utils/tasks/listTable.ts [hunk 5]
```diff
@@ -28,6 +39,49 @@ const getTaskType = (task: Task): string => {
...
+
+const getPlacementApplicationDecisionOutcome = (outcome: PlacementApplicationDecision): string => {
+  return placementApplicationDecisionTranslations[outcome]
+}
+
+const getPlacementRequestTaskOutcome = (outcome: PlacementRequestTaskOutcome): string => {
+  return placementRequestTaskOutcomeTranslations[outcome]
+}
+
+export const getDecisionOutcome = (task: Task): string => {
```
Comment (richpjames) [inline server/utils/tasks/listTable.ts:71]: In the After screenshot it looks like this function is returning `undefined` 
image

File: server/utils/tasks/listTable.ts [hunk 6]
```diff
@@ -93,13 +147,44 @@ const unallocatedTableRows = (tasks: Array<Task>): Array<TableRow> => {
...
+
+const completedTableRows = (tasks: Array<Task>): Array<TableRow> => {
+  const rows = [] as Array<TableRow>
+
+  tasks.forEach(task => {
+    rows.push([
+      nameAnchorCell(task),
+      completedAtDateCell(task),
+      completedByCell(task),
+      taskTypeCell(task),
```
Comment (richpjames) [inline server/utils/tasks/listTable.ts:170]: From the 'After' screenshot it looks like the `Task type` column is empty, it would be good to understand why

**Comment by vubalasu** [comment]
@richpjames pushed please check

File: server/utils/tasks/listTable.test.ts [hunk 7]
```diff
@@ -331,14 +376,88 @@ describe('table', () => {
...
+          completedAtDateCell(task),
+          completedByCell(task),
+          taskTypeCell(task),
+          decisionCell(task),
+        ],
+      ])
+    })
+  })
+
+  describe('completedTableHeaders', () => {
```
Comment (richpjames) [inline server/utils/tasks/listTable.test.ts:425]: This should be `tasksTableHeader`

**Review by richpjames — APPROVED** [review]

**Review by pezholio — APPROVED** [review]

**Pull request merged**